apps/emqx_cluster_link/src/emqx_cluster_link.erl

```diff
@@ -0,0 +1,97 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_cluster_link).
+
+-behaviour(emqx_external_broker).
+
+-export([
+    maybe_add_route/1,
+    maybe_delete_route/1,
+    forward/2,
+    is_relevant_route/2
+]).
+
+%% emqx hooks
+-export([
+    put_hook/0,
+    delete_hook/0,
+    on_message_publish/1
+]).
+
+-include("emqx_cluster_link.hrl").
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("emqx/include/emqx_hooks.hrl").
+-include_lib("emqx/include/logger.hrl").
+
+%%--------------------------------------------------------------------
+%% emqx_external_broker API
+%%--------------------------------------------------------------------
+
+maybe_add_route(Topic) ->
+    emqx_cluster_link_coordinator:route_op(<<"add">>, Topic).
+
+maybe_delete_route(Topic) ->
+    emqx_cluster_link_coordinator:route_op(<<"delete">>, Topic).
+
+forward(ExternalDest, Delivery) ->
+    emqx_cluster_link_mqtt:forward(ExternalDest, Delivery).
+
+%% TODO: a better function name, it should be still quite abstract to hide external details from emqx app..
+%% Do not forward any external messages to other links. Only forward local messages to all the relevant links.
+is_relevant_route({external, {link, _}}, #message{extra = #{sender_link := _}}) ->
+    false;
+is_relevant_route(_ExtDest, _Msg) ->
+    true.
+
+%%--------------------------------------------------------------------
+%% EMQX Hooks
+%%--------------------------------------------------------------------
+
+on_message_publish(#message{topic = <<?ROUTE_TOPIC_PREFIX, ClusterName/binary>>, payload = Payload}) ->
+    _ =
+        case emqx_cluster_link_mqtt:decode_route_op(Payload) of
+            {add, Topics} when is_list(Topics) ->
+                add_routes(Topics, ClusterName);
+            {add, Topic} ->
+                emqx_router_syncer:push(add, Topic, ?DEST(ClusterName), #{});
+            {delete, _} ->
+                %% Not implemented yet
+                ok
+        end,
+    ok;
```

Comment:
return `{stop, []}` here to terminate the hook chain.
NOTE: `[]` does not seem to be acceptable by the current hook caller, so we'll need to add a new case clause to support `[]` and `[Message | Batch]`.

(batching is not supported for now, but we could use this interface for future extension. For now, `[]` is to terminate publish, and `#message{...}` when forward is accepted.

Reply:
Done